spec/controllers/reports_controller_spec.rb

```diff
@@ -0,0 +1,334 @@
+  let!(:report_4) {FactoryBot.create :report, id: 4, from_user_id: user_singe.id, to_user_id: user_manager.id, department_id: department.id}
+
+  describe "GET #index" do
+    it_behaves_like "not logged for get method", "index"
+
+    context "when user logged" do
+      context "with normal user" do
+        before do
+          get :index
+        end
+
+        context "when has reports" do
+          it "with reports for normal user" do
+            log_in user
+            params = {
+              department_id: department.id,
+            }
+            get :index, params: params
+            expect(assigns(:reports).pluck(:id)).to eq([report_1.id, report_2.id, report_3.id])
+          end
+
+          it "with filter reports" do
+            log_in user
+            params = {
+              department_id: department.id,
+              filter: {
+                id: report_1.id,
+                department: nil,
+                name: nil,
+                date_created: nil,
+                date_reported: nil,
+                status: nil
+              }
+            }
+            get :index, params: params
+            expect(assigns(:reports).pluck(:id)).to eq([report_1.id])
+          end
+
+          it "with reports for manager" do
+            log_in user_manager
+            params = {department_id: department.id}
+            get :index, params: params
+            expect(assigns(:reports).pluck(:id)).to eq([report_1.id, report_2.id, report_3.id, report_4.id])
+          end
+        end
+
+        it "render index" do
+          log_in user
+          params = {department_id: department.id}
+          get :index, params: params
+          expect(response).to render_template :index
+        end
+      end
+
+      context "with invalid user login" do
+        before {log_in user_singe}
+        it_behaves_like "find a department", :index
+        it_behaves_like "find a relationship", :index
+      end
+    end
+  end
+
+  describe "GET show/:id" do
+    let!(:comment_1) {FactoryBot.create :comment, user_id: user.id, report_id: report_1.id}
+    let!(:comment_2) {FactoryBot.create :comment, user_id: user.id, report_id: report_1.id}
+    let!(:comment_3) {FactoryBot.create :comment, user_id: user.id, report_id: report_1.id}
+
+    it_behaves_like "not logged for other method" do
+      before do
+        get :show, params: {id: report_1.id}
+      end
+    end
+
+    context "when user logged" do
+      before {log_in user}
+
+      it_behaves_like "find a report", :show
+
+      context "when found" do
+        it "show comments latest" do
+          get :show, params: {id: report_1.id, department_id: department.id}
+          expect(assigns(:comments).pluck(:id)).to eq([comment_1.id,comment_2.id, comment_3.id])
+        end
+      end
+
+      it "comment should constructor" do
+        get :show, params: {id: report_1.id, department_id: department.id}
+        expect(assigns(:comment)).to_not eq nil
```

Comment:
mỗi 1 example - 1 khối it chỉ nên có 1 expected thôi e nhé

Reply:
e đọc chuẩn viết test áp dụng tại cty mình nhé
https://github.com/framgia/coding-standards/blob/master/vn/rails/test.md